### PR TITLE
Preparation for pages

### PR DESCRIPTION
**Pull request by kof**
Preparing the database to build pages feature #54 

## Before requesting a review

- [x] if the PR is WIP - use draft mode
- [x] made a self-review
- [x] added inline comments where things may be not obvious (the "why", not "what")
- [x] what kind of review is needed?
  - [x] conceptual
  - [ ] detailed
  - [ ] with testing

## Test cases

- [ ] step by step interaction description and what is expected to happen

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio-designer/blob/main/apps/designer/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `designer/env-check.js` if mandatory

### CODE REVIEW

File: apps/designer/app/shared/db/build.server.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+const Page = z.object({
+  id: z.string(),
+  name: z.string(),
+  path: z.string(),
+  title: z.string(),
+  meta: z.record(z.string(), z.string()),
+});
+const Pages = z.array(Page);
+type Pages = z.infer<typeof Pages>;
```
Comment (rpominov) [inline apps/designer/app/shared/db/build.server.ts:12]: In Webflow they have Folder as separate type from Page. It has no content, and only `name` and `slug` as properties. Should we do the same? Or do we want folders to also act as pages?

As a user it feels kinda like I'm forced to create some content for pages that I don't need. For example, if I'm building a blog I might want following pages:

 - `/` — list of posts
 - `/posts/first-post`
 - `/posts/second-post`

And I don't want anything at `/posts/`

Comment (rpominov) [reply apps/designer/app/shared/db/build.server.ts:12]: Might be a good idea to have `Homepage` separate from other pages:

```ts
type Pages = {
  homepage: Page;
  pages: array<Folder | Page>;
}
```

This will enforce 2 requirements:

 - Homepage must exist
 - Homepage must not be nested in any folder

Comment (kof) [reply apps/designer/app/shared/db/build.server.ts:12]: good idea

Comment (kof) [reply apps/designer/app/shared/db/build.server.ts:12]: > In Webflow they have Folder as separate type from Page. It has no content, and only name and slug as properties. Should we do the same? Or do we want folders to also act as pages?


@taylornowotny do you think this would be useful? I was also thinking about it and couldn't think of a use case

Comment (kof) [reply apps/designer/app/shared/db/build.server.ts:12]: > As a user it feels kinda like I'm forced to create some content for pages that I don't need. For example, if I'm building a blog I might want following pages:
> 
> / — list of posts
> /posts/first-post
> /posts/second-post
> And I don't want anything at /posts/

We failed to mention the "path" we let user enter is not a slug, user can enter a full path, folder doesn't create a path in our case, that's why having folders in our case is actually like folders: If you want a list you can still use /posts in the posts page

Comment (kof) [reply apps/designer/app/shared/db/build.server.ts:12]: One thing that is a little weird is that you would have a page called "posts" on the same level as the pages for each specific post.

Comment (rpominov) [reply apps/designer/app/shared/db/build.server.ts:12]: > We failed to mention the "path" we let user enter is not a slug, user can enter a full path

Hm, why do we need folders at all then? What if we implement pages as a flat list, at least for the first release? Would safe us a lot of time.

Comment (rpominov) [reply apps/designer/app/shared/db/build.server.ts:12]: Although if we later decide to control pages' URLs via nesting in folders, it will be hard to migrate to that from arbitrary paths.

Comment (kof) [reply apps/designer/app/shared/db/build.server.ts:12]: I have been back and forth on this too. Folders adding to the path force a certain URL structure you might not want to have, this is the feedback I got from @taylornowotny .

At the same time folders are good for organizing pages into categories of some sort when you have a lot of them. 

I was wondering if we could go with flat list of pages for beta too and wasn't sure if its going to save us a lot of time if we go with the same datastructure in the database.

The thing is if we want to leverage the patches, we would still need to update a single object. Potentially we could be updating one page at a time, but never the entire pages object. Then we have to ensure the right sort order. Having a single data structure for all pages ensures sort order automatically.

If we go with a flat list now, how hard will be adding folders in the future?

Thoughts?

Comment (kof) [reply apps/designer/app/shared/db/build.server.ts:12]: I don't think we truly need folders for beta release, but I was thinking its not gonna save us much time if we don't have them

Comment (kof) [reply apps/designer/app/shared/db/build.server.ts:12]: Regarding folder enforcing a path: I guess we could even make it an option. By default folder would have no path, but if added, it gets added automatically to all pages inside that folder, wdyt @taylornowotny ?

Comment (kof) [reply apps/designer/app/shared/db/build.server.ts:12]: Another thought: what if pages and folders are the same? If user doesn't want to use folder as a page, they will simply not use it. Nothing forces them to use that page, they are linking the pages anyways manually. So if its not linked its not used ... ? @taylornowotny

Comment (rpominov) [reply apps/designer/app/shared/db/build.server.ts:12]: > The thing is if we want to leverage the patches, we would still need to update a single object. Potentially we could be updating one page at a time, but never the entire pages object. Then we have to ensure the right sort order. Having a single data structure for all pages ensures sort order automatically.

Yeah, I'd still kept them in one JSON object event if they a flat list.

> If we go with a flat list now, how hard will be adding folders in the future?

A bit hard to tell, but I feel like if folders are used only for organizing, then there won't be much additional work compared to doing it now: `work to implement now with folders ~= now without folders + folders later`.

And this will allow us to think more about this. Maybe we'll decide to have tags instead of folders, or something like that.

Comment (kof) [reply apps/designer/app/shared/db/build.server.ts:12]: yeah, the question is if it saves us much time, in that case, but yeah, thinking through all those questions properly would be highly benefitial

Comment (rpominov) [reply apps/designer/app/shared/db/build.server.ts:12]: >  the question is if it saves us much time

Hard to tell. I think from a day to a week

Comment (kof) [reply apps/designer/app/shared/db/build.server.ts:12]: Ok lets first figure out without pages and then see if we can bring them in within a few days right away. We would still use the same tree component for drag&drop etc, so supporting dragging between folders should be easy

File: apps/designer/app/designer/designer.tsx [hunk 2]
```diff
@@ -219,7 +219,7 @@ const NavigatorPanel = ({ publish, isPreviewMode }: NavigatorPanelProps) => {
 
 type DesignerProps = {
   config: Config;
-  project: Project;
+  project: db.project.Project;
```
Comment (kof) [inline apps/designer/app/designer/designer.tsx:222]: btw this is a total workaround, we need to have a way to get the augmented Project type, which is currently not importable in a good way, we need a project package I think

File: apps/designer/app/shared/db/build.server.ts (hunk 1 above)
Comment (taylornowotny) [reply apps/designer/app/shared/db/build.server.ts:12]: Re: folders determining the path - I don't think is something we need to include as an option. Maybe some users would want to use folders to organize paths (like "/folder/page") but since they have full control over the path on every page anyway, they can set that up manually.

Comment (taylornowotny) [reply apps/designer/app/shared/db/build.server.ts:12]: Folders are meant to be purely organizational

Comment (taylornowotny) [reply apps/designer/app/shared/db/build.server.ts:12]: We could use tags instead of folders, but I think tags would be an inferior organizational tool in this case for the following reasons:

- folders are better for visual organization in the pages list because of nesting. Users can nest pages to hide them from view while they focus on other pages. With tags there is no nesting
- both tags and folders are good for grouping pages, but there's not currently a UI for tags so that would have to be designed and built
- tags are best used with sorting and filtering controls, which are best used for large sets. An average website might have a dozen pages or less, so the full utility of tags is wasted on such a small list.

File: apps/designer/app/designer/designer.tsx (hunk 2 above)
Comment (rpominov) [reply apps/designer/app/designer/designer.tsx:228]: Might be better to have a single package that works with database. Its API would be based on the augmented types, and as a user of this package I would know nothing about `postgres` or maybe even `prisma`. 

But it should handle all database objects. Otherwise I feel like we'll end up in a dependency hell and might even have to deal with circular NPM dependencies which I don't think possible.

Comment (kof) [reply apps/designer/app/designer/designer.tsx:228]: The problem with this centralized approach is that it will end up depending on a LOT of other types across the application.

There is a another way - shared functionality between features always goes into a shared folder that is named based on its function/scope.

So for example there would be a shared/project folder that has everything about the project including augmented database types and utilities.

Comment (kof) [reply apps/designer/app/designer/designer.tsx:228]: The way it is right now is not great, because we tried initially to centralize database stuff in a shared/db folder and this is currently the result.

Comment (kof) [reply apps/designer/app/designer/designer.tsx:228]: Ultimately if we have it this way no feature will have to access prisma client directly, they will all work with shared/project shared/instance etc, which in return will call into prisma and augment data and types

File: apps/designer/app/shared/db/build.server.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  path: z.string(),
+  title: z.string(),
+  meta: z.record(z.string(), z.string()),
+  treeId: z.string(),
+});
+
+type Page = z.infer<typeof Page>;
+
+const Pages: z.ZodType<{ homePage: Page; pages: Array<Page> }> = z.object({
+  homePage: Page,
```
Comment (kof) [inline apps/designer/app/shared/db/build.server.ts:16]: why not make it an id? 
```suggestion
  home: Page['id'],
```

Comment (kof) [reply apps/designer/app/shared/db/build.server.ts:16]: ah pages won't have the home page? I thought pages will have all the pages, and one of them is home

Comment (rpominov) [reply apps/designer/app/shared/db/build.server.ts:16]: I'm a bit obsessed with the idea of creating data-structures that make invalid states impossible to represent https://www.youtube.com/watch?v=IcgmSRJHu_8

Here, if we go with `id`, it may happen that the pages array will not have a page with that `id`.

But I think maybe we should just make it a simple array after all, with a `boolean` property `Page.isHome`. Need to see how a separate `homePage` complicates or doesn't complicate the code.

Comment (rpominov) [reply apps/designer/app/shared/db/build.server.ts:16]: Will probably go with a simple array + checks for all requirements for home page in the zod schema.

Comment (kof) [reply apps/designer/app/shared/db/build.server.ts:16]: > I'm a bit obsessed with the idea of creating data-structures that make invalid states impossible to represent

me too, definitely

> Here, if we go with id, it may happen that the pages array will not have a page with that id.

could be ensured by the logic, but yeah its less robust

Comment (kof) [reply apps/designer/app/shared/db/build.server.ts:16]: if you go with an isHome you have to ensure only one has it, if you go with a home: id - you only need to ensure the id still exists in the tree, so user can't delete home page without setting to home some other page

**Comment by rpominov** [comment]
Something we decided in a chat about relationships of build an projects:

 - Project always refers to the current dev build
 - Project may refer to a prod build 
 - Build refers to a project to allow us to get history of all builds of a project
 - We should not assume that prod or dev build are chronologically most recent

File: packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      );
+
+      if (project === undefined) {
+        continue;
+      }
+
+      builds.push({
+        pages: JSON.stringify({
+          homePage: {
+            id: uuidv4(),
```
Comment (rpominov) [inline packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:27]: For `Page.id` I think it's better to use UUID instead of `bson-objectid` that we use for `Instance.id`. Main reason is that UUID can be easily created in an SQL query if we ever need to create a Page using just SQL.

File: packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import { PrismaClient, Prisma } from "./client";
+import { v4 as uuidv4 } from "uuid";
```
Comment (kof) [inline packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:2]: could just map it to uuid for simplicity

File: packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    const projects = await prisma.project.findMany();
+
+    const builds: Prisma.BuildCreateManyInput[] = [];
+
+    for (const tree of trees) {
+      const project = projects.find(
+        (project) =>
+          project.devTreeId === tree.id ||
+          project.prodTreeId === tree.id ||
+          JSON.parse(project.prodTreeIdHistory).includes(tree.id)
```
Comment (kof) [inline packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:17]: I would also run it over zod schema to be sure

Comment (kof) [reply packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:17]: I mean in all cases where we JSON.parse or stringify

File: packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+          },
+          pages: [],
+        }),
+        isDev: project.devTreeId === tree.id,
+        isProd: project.prodTreeId === tree.id,
+        projectId: project.id,
+      });
+    }
+
+    await prisma.build.createMany({ data: builds });
```
Comment (kof) [inline packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:42]: did you mean to run this synchronously? or could all of these run in parallel?

File: packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts (hunk 6 above)
Comment (rpominov) [reply packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:17]: This is a bit tricky. All code that is used in a migration has to be immutable to changes that we make to the rest of the project. We can't just import something. We would have to copy everything we want to use into the migration folder.

File: packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts (hunk 7 above)
Comment (rpominov) [reply packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:42]: Not sure I understand the question. If you're asking why I use `createMany` instead of `create`, that's because with `createMany` Prisma should make just one SQL query (it may split in to several batches, but still less that a query for each row)

Comment (rpominov) [reply packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:42]: Doesn't matter for the amount of data we a're working with. Just want to set a good example.

File: packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts (hunk 6 above)
Comment (kof) [reply packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:17]: oh that's an interesting point, definitely need to document that! maybe even add to each migration file as a comment and in a template, because this is SUPER non-obvious

Comment (kof) [reply packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:17]: I guess we could copy the schema we need to the local modules, just like the client

File: packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts (hunk 7 above)
Comment (kof) [reply packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:42]: what I mean is  that this current implementation will run one query after another in that for of loop, but it could run in parallel, I see 2 options:

1. using map, return promises from queries without awaiting, then Promise.all - this will send many queries all in parallel
2. use transaction API https://www.prisma.io/docs/concepts/components/prisma-client/transactions#the-transaction-api, not sure if prisma does there anything different than in 1 (I recall they can optimize that, not sure though)

File: packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts (hunk 6 above)
Comment (rpominov) [reply packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:17]: > maybe even add to each migration file as a comment and in a template

Good idea! Will add to the template

> I guess we could copy the schema we need to the local modules, just like the client

Something to keep in mind, but for now will just copy Zod definitions to the migration file manually

File: packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts (hunk 7 above)
Comment (rpominov) [reply packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:42]: > one query after another in that "for of" loop

There're no queries in the loop. Entire migration is 3 queries.

> using map, return promises from queries without awaiting, then Promise.all - this will send many queries all in parallel

This will work only if Prisma uses several connections to the database AND if we won't wrap the entire migration into a transaction (a transaction can't be run over multiple connections afaik)

> use transaction API https://www.prisma.io/docs/concepts/components/prisma-client/transactions#the-transaction-api, not sure if prisma does there anything different than in 1 (I recall they can optimize that, not sure though)

Same. Prisma can't run queries in a transaction in parallel, that's just a limitation on Postgres level

Comment (kof) [reply packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:42]: Either I am missing it or you are: we are running those queries from some server. That server has a latency to the database server.

When we send the query and await for a response, we are waiting for execution + latency to receive the results.
When we parallelize like I said - queries are send all at once, we are waiting for all responses, no matter which one comes back first.

So I am optimizing away the latency between the database server and the node server.

Comment (kof) [reply packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:42]: If it's just 3 queries it may not matter, but in a complex migration that could be many queries

File: packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts (hunk 6 above)
Comment (kof) [reply packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:17]: that's what I meant, zod schema definitions, copied, could be separate files next to the migration

File: packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts (hunk 7 above)
Comment (kof) [reply packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts:42]: you are right, I misread when this query happens because of indentations, I thought it's inside the for-of loop but its after that and you are already using transaction, so I guess prisma will send all queries together

File: apps/designer/app/shared/db/build.server.ts (hunk 3 above)
Comment (rpominov) [reply apps/designer/app/shared/db/build.server.ts]: Keeping `homePage: Page` for now, may reconsider later

**Comment by kof** [comment]
@rpominov "Project may refer to a prod build"

What do you mean by this? Just to clarify what I thought initially is that project is kinda always in dev mode, so refers a dev build. When we publish, we clone that build and reference it as a prod build. So prod build is never modified by accident, it's always secure and we always operate on the dev build. 

I guess you already know that, just making sure

**Comment by rpominov** [comment]
@kof Yeah, we're on the same page I think. Sorry if I described it in a confusing way. Was trying to describe it from the perspective of how it's laid out in the database.

**Comment by rpominov** [comment]
Please ignore the current solution for now, btw. It's still in a flux. I need to finish updating the code that works with data to see which way is better. Let's discuss after PR is ready for review.

**Comment by rpominov** [comment]
@kof This is ready for review (can't request as you're the author 😅 )

File: apps/designer/app/canvas/shared/sync/use-sync.ts [hunk 8]
```diff
@@ -17,7 +17,7 @@ export const useSync = ({ project }: { project: Project }) => {
         method: "post",
         body: JSON.stringify({
           transactions: entries,
-          treeId: project.devTreeId,
+          treeId: project.devBuild?.pages.homePage.treeId,
```
Comment (kof) [inline apps/designer/app/canvas/shared/sync/use-sync.ts:20]: this will depend on the currently selected page, right? What's the plan, is this temporary and needs a todo?

**Comment by andrasbacsai** [comment]
Merged main and resolved all conflicts, requested by @kof  (there was a lot 😅). 

@rpominov Please review when you are back.

File: apps/designer/app/routes/rest/breakpoints.$projectId.ts [hunk 9]
```diff
@@ -10,19 +10,21 @@ export const loader: LoaderFunction = async ({
...
+
+    const prodBuild = await db.build.loadProdByProjectId(params.projectId);
+
+    if (prodBuild === undefined) {
       throw new Error(
         `Project ${params.projectId} needs to be published first`
       );
     }
-    const data = await db.breakpoints.load(project.prodTreeId);
+    const data = await db.breakpoints.load(prodBuild.pages.homePage.treeId);
```
Comment (kof) [inline apps/designer/app/routes/rest/breakpoints.$projectId.ts:24]: breakpoints is per build, not per page, we need to transform it to use build id

File: apps/designer/app/routes/rest/props.$projectId.ts [hunk 10]
```diff
@@ -9,8 +9,18 @@ export const loader: LoaderFunction = async ({
   params,
 }): Promise<Array<InstanceProps> | ErrorData> => {
   try {
-    const project = await db.project.loadById(params.projectId);
-    return await db.props.loadByProject(project, "production");
+    if (params.projectId === undefined) {
+      throw new Error(`Project ID required`);
+    }
+
+    const prodBuild = await db.build.loadProdByProjectId(params.projectId);
```
Comment (kof) [inline apps/designer/app/routes/rest/props.$projectId.ts:16]: wouldn't it read more nicely if we had build.loadByProjectId(projectId, 'prod') ?

File: apps/designer/app/shared/db/misc.server.ts [hunk 11]
```diff
@@ -23,24 +23,16 @@ export const publish = async ({
...
-  });
-  const prodTreeIdHistory = project.prodTreeIdHistory;
-  if (project.prodTreeId) {
-    prodTreeIdHistory.push(project.prodTreeId);
+  // should not happen
+  if (project.devBuild == undefined) {
+    throw new Error(`project.devBuild is not loaded`);
   }
+
+  await db.build.createProd(project.devBuild, project.id);
```
Comment (kof) [inline apps/designer/app/shared/db/misc.server.ts:31]: same suggestion here `build.create(project.id, project.devBuild, "prod")`

File: packages/prisma-client/migrations-cli/commands.ts [hunk 12]
```diff
@@ -129,6 +129,26 @@ const ensureLastPending = async (migrationsName: string) => {
   }
 };
 
+const isNoopSql = (sqlScript: string) => {
```
Comment (kof) [inline packages/prisma-client/migrations-cli/commands.ts:132]: you probably wanna split these into a separate pr and we can merge it quickly

File: packages/project/src/db/build.ts [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+      treeId: tree.id,
+    },
+    pages: [],
+  });
+};
+
+export const createDev = async (baseBuild?: Build) => {
+  if (baseBuild === undefined) {
+    const pages = await createPages();
+    return prisma.build.create({
```
Comment (kof) [inline packages/project/src/db/build.ts:85]: not awaiting?

File: packages/project/src/db/build.ts [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+      data: { pages: JSON.stringify(pages) },
+    });
+  } else {
+    const pages = await clonePages(baseBuild.pages);
+    return prisma.build.create({
```
Comment (kof) [inline packages/project/src/db/build.ts:90]: same

File: packages/project/src/db/project.ts [hunk 15]
```diff
@@ -119,43 +121,20 @@ export const clone = async (clonableDomain: string, userId: string) => {
...
-  ]);
-
-  const parsedProject = parseProject(project);
-  if (parsedProject === null) {
-    throw new Error(`Not found project "${clonableDomain}"`);
-  }
-  return parsedProject;
+  const devBuild = await db.build.createDev(prodBuild);
+
+  return create({
```
Comment (kof) [inline packages/project/src/db/project.ts:133]: await?

File: packages/project/src/db/types.ts [hunk 16]
```diff
@@ -1,9 +1,33 @@
...
+import {
+  Project as BaseProject,
+  Build as DbBuild,
+} from "@webstudio-is/prisma-client";
 import type { Asset } from "@webstudio-is/asset-uploader";
-const TreeHistorySchema = z.array(z.string());
 
-export type Project = Omit<BaseProject, "prodTreeIdHistory" | "assets"> & {
-  prodTreeIdHistory: z.infer<typeof TreeHistorySchema>;
+// Project
```
Comment (kof) [inline packages/project/src/db/types.ts:8]: useful comment :)

**Comment by rpominov** [comment]
Thanks @andrasbacsai !

File: packages/project/src/db/build.ts (hunk 13 above)
Comment (rpominov) [reply packages/project/src/db/build.ts:85]: Not sure I understand

```
const fn = async () => {
  return promise
}
```

Is equivalent to

```
const fn = async () => {
  const result = await promise
  return result
}
```

Or do we consider this a bad practice?

Comment (kof) [reply packages/project/src/db/build.ts:85]: oh is it? I didn't think this through

File: apps/designer/app/canvas/shared/sync/use-sync.ts (hunk 8 above)
Comment (rpominov) [reply apps/designer/app/canvas/shared/sync/use-sync.ts:20]: Yeah, temporary. We wanted to only change schema in this PR, and make minimal changes to the reset of the code, just to make it work.

Will add todos.

Comment (kof) [reply apps/designer/app/canvas/shared/sync/use-sync.ts:20]: gotcha, as you wish, as long as we can merge it and potentially release it, all good

**Comment by andrasbacsai** [comment]
Merged the main again, because there was some problem with `projects` package.

File: packages/prisma-client/migrations-cli/commands.ts (hunk 12 above)
Comment (rpominov) [reply packages/prisma-client/migrations-cli/commands.ts:132]: Sorry for mixing this up. I'll do better next time, but I'd safe time on splitting and merge this as is

**Review by rpominov — APPROVED** [review]

**Comment by rpominov** [comment]
@kof Ready for review

File: packages/prisma-client/prisma/migrations/template.txt [hunk 17]
```diff
@@ -10,7 +10,10 @@ import { PrismaClient } from "./client";
 //   It's better to copy it to the migration directory.
 
 export default () => {
-  const client = new PrismaClient();
+  const client = new PrismaClient({
+    // Uncomment to see the queries in console as the migration runs
+    // log: ["query", "info", "warn", "error"],
```
Comment (kof) [inline packages/prisma-client/prisma/migrations/template.txt:15]: you can also do thi https://www.prisma.io/docs/concepts/components/prisma-client/debugging

**Review by kof — COMMENTED** [review]
lgtm, pelase check the preview env too

File: packages/prisma-client/prisma/migrations/template.txt (hunk 17 above)
Comment (rpominov) [reply packages/prisma-client/prisma/migrations/template.txt:15]: I tried this, but it doesn't output queries in SQL format, only in some prisma internal format

**Comment by rpominov** [comment]
Preview won't work because migrations are not applied
